### ISSUE TITLE
The iOS mobile app version 24.4.0 cannot handle websockets using TLS v1.3 (only TLS v1.2)

### ISSUE THREAD

**Issue by tomas-sve** [jitsi/jitsi-meet#15044]
### What happened?

The iOS mobile app 24.4.0 (current version) cannot handle websockets using TLS v1.3 (only TLS v1.2).

If you only allow TLS v1.3 on your jitsi meet server you get immediately disconnected when you try to join a meeting using the latest iOS app. In the debug log you can see that the TLS negotiation for the websocket connection fails with **errSSLPeerProtocolVersion**.

The Android app 24.4.0 can still use websockets with TLS versions 1.2 and 1.3 and so could earlier versions of the iOS app (version 22.5.1 is confirmed to work with TLS v1.3).

### Platform

- [ ] Chrome (or Chromium based)
- [ ] Firefox
- [ ] Safari
- [ ] Other desktop browser
- [ ] Android browser
- [ ] iOS browser
- [ ] Electron app
- [ ] Android mobile app
- [X] iOS mobile app
- [ ] Custom app using a mobile SDK

### Browser / app / sdk version

Version 24.4.0, SDK version 10.1.0

### Relevant log output

```shell
Output from iPhone 13 and the Jitsi Meet App Version 24.4.0:

Aug 22 14:28:36 jitsi-meet(CFNetwork)[594] <Notice>: TCP Conn 0x30338e580 started
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Notice>: TCP Conn 0x30338e580 event 1. err: 0
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Notice>: TCP Conn 0x30338e580 complete. fd: 58, err: 0
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Notice>: TCP Conn 0x30338e580 starting SSL negotiation
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Notice>: CFNetwork SSLHandshake failed (-9836)
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Error>: TCP Conn 0x30338e580 SSLHandshake failed (-9836)
Aug 22 14:28:37 jitsi-meet(CFNetwork)[594] <Notice>: TCP Conn 0x30338e580 canceled
Aug 22 14:28:37 jitsi-meet(JitsiMeetSDK)[594] <Error>: [modules/xmpp/strophe.util.js] Strophe: Websocket error {"isTrusted":false,"message":"The operation couldn\M-b\M^@\M^Yt be completed. (OSStatus error -9836.)"} Websocket error {"isTrusted":false,"message":"The operation couldn\M-b\M^@\M^Yt be completed. (OSStatus error -9836.)"}
Aug 22 14:28:37 jitsi-meet(JitsiMeetSDK)[594] <Error>: [modules/xmpp/strophe.util.js] Strophe: Websocket closed unexcectedly Websocket closed unexcectedly
Aug 22 14:28:37 jitsi-meet(JitsiMeetSDK)[594] <Error>: [features/base/conference] JWT error: The WebSocket connection could not be established or was disconnected.
```

### Reproducibility

- [ ] The problem is reproducible on meet.jit.si

### More details?

The return code -9836 means errSSLPeerProtocolVersion and if I change the nginx config of my jitsi meet server to allow both TLSv1.2 and TLSv1.3 instead of just TLSv1.3 then it starts working.

The problem doesn't show on meet.jit.si because it allows TLSv1.2.

iOS version 17.6.1 was used for testing.

**Comment by saghul** [comment]
Thanks for the report!



> The Android app 24.4.0 can still use websockets with TLS versions 1.2 and 1.3 and so could earlier versions of the iOS app (version 22.5.1 is confirmed to work with TLS v1.3).

This is surprising! We have updated the RN version, but I wouldn't have guessed such a change was made. We'll investigate.

**Comment by randomthingsandstuff** [comment]
looking to see if this can be reopened. 24.6.0 affected as well. kind of a pain to debug too 😄

**Comment by saghul** [comment]
Any chance you can give 25.0.0 a try?

**Comment by randomthingsandstuff** [comment]
no luck still on 25.0.0.

**Comment by saghul** [comment]
Looks like the problem is in the library RN uses on iOS, SocketRocket: https://github.com/facebookincubator/SocketRocket/blob/a9872b7a14a990a92266e021cfaece52d79b8609/SocketRocket/SRSecurityPolicy.m#L60

I decided to try my luck and sent a PR: https://github.com/facebookincubator/SocketRocket/pull/701

**Comment by mf-conference** [comment]
Is there any update about this?

**Comment by saghul** [comment]
Yes, I applied the patch to our build already, it should be working now.

**Comment by mf-conference** [comment]
> Yes, I applied the patch to our build already, it should be working now.

Yeah patch seems correct for 1.1,1.2 & 1.3 However it doesn't enforce to 1.3 only right?

**Comment by saghul** [comment]
It should enforce TLS >= 1.2, so if you only offer 1.3 it should work.

**Comment by arslanarso** [comment]
I’m connecting to a server that only supports TLSv1.3. I’ve already applied this patch, but according to the logs, the client still connects using TLSv1.2 instead of TLSv1.3. Any idea why this might be happening?

**Comment by arslanarso** [comment]
I’ve done some research, and I found this [Stack Overflow thread](https://stackoverflow.com/questions/77529907/tlsv1-3-not-supporting-in-ios-17-ios)
 mentioning that even on iOS 17, TLSv1.3 is not fully supported.
It seems this issue originates from the SocketRocket library itself, where the same TLSv1.3 limitation has been reported.

Also, according to Apple’s official documentation ([here](https://support.apple.com/tr-tr/guide/security/sec100a75d12/web)), the SecureTransport APIs do not support TLS 1.3.

I tested this patch on an iOS 18 device, but when connecting to a server that only accepts TLSv1.3, the client still sends a TLS 1.2 ClientHello. Since the handshake can’t complete, the connection fails.

It seems the issue is related to SocketRocket, as this limitation has also been observed there. I also found [this open issue](https://github.com/facebookincubator/SocketRocket/issues/672)
 on GitHub, but there hasn’t been any response so far.

I tried several different ways to enforce TLS 1.3 or set the minimum TLS version to 1.3, but it always falls back to TLS 1.2 during the handshake.

Is there any known workaround or plan to make TLS 1.3 work with iOS clients? Or do we have to keep supporting TLS 1.2 for compatibility?

@saghul

**Comment by saghul** [comment]
Thanks for giving it a try! If my change wasn't enough and the Apple provided APIs are not sufficient I'm afraid big changes would be required in SocketRocket, which is not something we are ready to tackle.

I guess you need to keep supporting TLS 1.2.

**Comment by GaRUi** [comment]
NSStream API is will not be updated and it does not support TLS 1.3 

https://developer.apple.com/documentation/technotes/tn3151-choosing-the-right-networking-api

> CFSocketStream was marked as to-be-deprecated in [2021] Apple will not enhance it to support new features. For example, Apple added TLS 1.3 support to Network framework in [2019], but has not added it to CFSocketStream.

> The TCP networking support in NSStream, most notably [getStreamsToHostWithName:port:inputStream:outputStream:], is layered on top of CFSocketStream and is on the same deprecation path.

So is there a way to change this in SocketRocket ? SocketRocket is build on top of NSStream. 
Maybe switch to Apple's WebSocket will be easier ?  https://developer.apple.com/documentation/foundation/urlsessionwebsockettask

**Comment by saghul** [comment]
> So is there a way to change this in SocketRocket ?

Not an easy one since it's a React Native dependency. We can patch it, but I suspect it would be quite an invasibvle patch.

We don't have plans to tackle this ourselves, but if you are so inclined, you are welcome to give it a try and we can float the patch.
